metric_map.go

```diff
@@ -0,0 +1,229 @@
+package gostatsd
+
+import (
+	"bytes"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// MetricMap is used for storing aggregated or consolidated Metric values.
+// The keys of each map are metric names.
+type MetricMap struct {
+	Counters Counters
+	Timers   Timers
+	Gauges   Gauges
+	Sets     Sets
+}
+
+func NewMetricMap() *MetricMap {
+	return &MetricMap{
+		Counters: Counters{},
+		Timers:   Timers{},
+		Gauges:   Gauges{},
+		Sets:     Sets{},
+	}
+}
+
+// Receive adds a single Metric to the MetricMap, and releases the Metric.
+func (mm *MetricMap) Receive(m *Metric, now time.Time) {
+	tagsKey := m.TagsKey
+	nowNano := Nanotime(now.UnixNano())
+
+	switch m.Type {
+	case COUNTER:
+		mm.receiveCounter(m, tagsKey, nowNano)
+	case GAUGE:
+		mm.receiveGauge(m, tagsKey, nowNano)
+	case TIMER:
+		mm.receiveTimer(m, tagsKey, nowNano)
+	case SET:
+		mm.receiveSet(m, tagsKey, nowNano)
+	default:
+		logrus.StandardLogger().Errorf("Unknown metric type %s for %s", m.Type, m.Name)
```

Comment:
Is there sanitisation on m.Type before we reach this point? I don't like the idea of emitting error logs on bad user input (bad statsd clients / unsupported types) but if we should've caught it before here that's another matter.